gbpservice/nfp/configurator/agents/agent_base.py

```diff
@@ -0,0 +1,181 @@
+from oslo_log import log as logging
+from gbpservice.nfp.configurator.lib import (
+    generic_config_constants as const)
+
+LOG = logging.getLogger(__name__)
+
+"""Implements base class for all service agents.
+
+Common methods for service agents are implemented in this class. Configurator
+module invokes these methods through the service agent's child class instance.
+
+"""
+
+
+class AgentBaseRPCManager(object):
+
+    def __init__(self, sc, conf):
+        self._sc = sc
+        self.conf = conf
+
+    def validate_request(self, sa_info_list, notification_data):
+        if (isinstance(sa_info_list, list) and
+                isinstance(notification_data, dict)):
+            return True
+        else:
+            return False
+
+    def forward_request(self, sa_info_list, notification_data):
+        """Forwards the RPC message from configurator to service agents.
+
+        Checks if the request message contains multiple data blobs. If multiple
+        data blobs are found, a batch event is generated otherwise a single
+        event.
+
+        :param sa_info_list: List of data blobs prepared by de-multiplexer
+        for service agents processing.
+        :param notification_data: Notification blobs prepared by the service
+        agents after processing requests blobs. Each request blob will have
+        a corresponding notification blob.
+
+        Returns: None
+
+        """
+
+        # In case of malformed input, send failure notification
+        if not self.validate_request(sa_info_list, notification_data):
+            # TODO: Need to send failure notification
+            return
+
+        # Multiple request data blobs needs batch processing. Send batch
+        # processing event or do direct processing of single request data blob
+        if (len(sa_info_list) > 1):
+            args_dict = {
+                'sa_info_list': sa_info_list,
+                'notification_data': notification_data
+            }
+            ev = self._sc.new_event(id='PROCESS_BATCH', data=args_dict)
+            self._sc.post_event(ev)
+        else:
+            sa_info_list[0]['context'].update(
+                {'notification_data': notification_data})
+            sa_info_list[0]['context'].update(
+                {'resource': sa_info_list[0]['resource']})
+            getattr(self, sa_info_list[0]['method'])(
+                sa_info_list[0]['context'],
+                **sa_info_list[0]['kwargs'])
+
+
+class AgentBaseEventHandler(object):
+
+    def __init__(self, sc, drivers, rpcmgr, nqueue):
+        self._sc = sc
+        self.drivers = drivers
+        self._rpcmgr = rpcmgr
+        self.nqueue = nqueue
+
+    def _notification(self, data):
```

Comment:
Discussed with Ahmed. Looks like the poll_event is overloaded to use for enqueue purpose also right now. He is making this method as queue_event(). Once that is available, this poll_event() call will become queue_event()